File: Practices/Regionals/f.java
import java.util.*;
import java.io.*;

public class f {
    public static long[][] memo;
    public static int[] mess;
    public static int[] clean;
    public static boolean[] parents;


    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int numDays = in.nextInt();
        int numParents = in.nextInt();

        mess = new int[numDays];
        clean = new int[numDays];
        parents = new boolean[numDays];
        
        for (int i = 0; i < mess.length; i++) {
            mess[i] = in.nextInt();
            clean[i] = in.nextInt();
        }

        for (int i = 0; i < numParents; i++)
            parents[in.nextInt() - 1] = true;
        
        memo = new long[numDays + 1][numDays + 1];
        for (long[] oned : memo)
            Arrays.fill(oned, -1);

        for (int i = 0; i <= numDays; i++) {
            long ans = getMin(numDays-1, i);
            if(ans <= (long)1e9) {   
                System.out.println(i);
                return;
            }
        }

        System.out.println(-1);
    }

    private static long getMin(int at, int numCleans) {
        
        if (numCleans < 0)
            return (long)1e10;

        if(at == -1)
            return 0;
    
        if (memo[at][numCleans] != -1)
            return memo[at][numCleans];
    
        long didnt = Math.max(0, getMin(at-1, numCleans) + mess[at]);
        long did = Math.max(0, getMin(at-1, numCleans-1) +  mess[at] - clean[at]);

        if (parents[at])
            if (did!=0 && didnt!=0)
                return memo[at][numCleans] = (long)1e10;
    
        return memo[at][numCleans] = Math.min(did, didnt);
    }
}

File: Practices/2023/5/f.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using vi = vector<ll>;
using vvi = vector<vi>;
using vvvi = vector<vvi>;
#define all(x) begin(x), end(x)
#define rall(x) rbegin(x), rend(x)

void solve() {
    int n, d;
    cin >> n >> d;
    ll BAD = 1e10;
    vector<ll> mess(n);
    vector<ll> clean(n);
    for(int i = 0; i < n; i++){
        cin >> mess[i];
        cin >> clean[i];
    }
    vector<bool> parents(n);
    while(d--) {
        int x;
        cin >> x, x--;
        parents[x] = true;
    }

    vvi memo(n, vi(n+1, -1));

    auto go = [&](int at, int numCleans, auto&& go) -> ll {
        if(numCleans < 0) return BAD;
        if(at == -1) return 0;

        auto& ans = memo[at][numCleans];
        if(ans!=-1){
            // cout << "memo!\n";
            return ans;
        } 

        ll cleanhere = max(0LL, go(at-1, numCleans - 1, go) + mess[at] - clean[at]);
        ll dontClean = max(0LL, go(at-1, numCleans, go) + mess[at]);

        ll minn = min(cleanhere, dontClean);
        if(parents[at] && minn != 0) return ans = BAD;

        return ans = minn;

    };

    int lo = 0;
    int hi = n;
    int ans = -1;
    while(lo <= hi){
        int mid = (lo + hi) >> 1;

        // if this many cleans could do it.
        ll mess = go(n-1, mid, go);
        // cout << mid << " " << mess << " " << BAD << "\n";

        if(mess < 1e9) {
            hi = (ans = mid) - 1;
        } else {
            lo = mid + 1;
        }
    }

    cout << ans << "\n";

}

int main(){
    cin.tie(0)->sync_with_stdio(0);
    cin.exceptions(cin.failbit);
    solve(); return 0;
}